audisp-tacplus.c

```diff
@@ -1,652 +1,575 @@
+        if(srv_fd < 0) {
+            syslog(LOG_WARNING, "connection to %s failed (%d) to send"
+                " accounting record: %m",
+                tac_ntop(tac_srv[srv_i].addr->ai_addr), srv_fd);
+            continue;
+        }
+        retval = send_acct_msg(srv_fd, type, user, tty, host, cmdmsg, task_id);
+        if(retval)
+            syslog(LOG_WARNING, "error sending accounting record to %s: %m",
+                tac_ntop(tac_srv[srv_i].addr->ai_addr));
+        close(srv_fd);
+        if(!retval) {
+            connected_ok = 1;
+            if(!(tacacs_ctrl & PAM_TAC_ACCT))
+                break; /* only send to first responding server */
+        }
+    }
+}
+
+/*
+ * encapsulate the field lookup, and rewind if needed,
+ * rather than repeating at each call.
+ */
+static const char *get_field(auparse_state_t *au, const char *field)
+{
+    const char *str;
+    if(!(str=auparse_find_field(au, field))) {
+        auparse_first_field(au);
+        if(!(str=auparse_find_field(au, field))) {
+            /* sometimes auparse_first_field() isn't enough, depending
+             * on earlier lookup order. */
+            auparse_first_record(au);
+            if(!(str=auparse_find_field(au, field)))
+                return NULL;
+        }
+    }
+    return str;
+}
+
+/* find an audit field, and return the value for numeric fields.
+ * return 1 if OK (field found and is numeric), otherwise 0.
+ * It is somewhat smart, in that it will try from current position in case code
+ * is written "knowing" field order; if not found will rewind and try again.
+ */
+static unsigned long
+get_auval(auparse_state_t *au, const char *field, int *val)
+{
+    int rv;
+
+    if(!get_field(au, field))
+        return 0;
+    rv = auparse_get_field_int(au);
+    if(rv == -1 && errno)
+        return 0;
+    *val = rv;
+    return 1;
+}
+
+
+/*
+ * Get the audit record for exec system calls, and send it off to the
+ * tacacs+ server.   Lookup the original tacacs username first.
+ * This just gets us the starts of commands, not the stop, which would
+ * require matching up the exit system call.   For now, don't bother.
+ * Maybe add some caching of usernames at some point.
+ * Both auid and sessionid have to be valid for us to do accounting.
+ * We don't bother with really long cmd names or really long arg lists,
+ * we stop at 240 characters, because the longest field tacacs+ can handle
+ * is 255 characters, and some of the accounting doesn't seem to work
+ * if right at full length.
+ */
+static void get_acct_record(auparse_state_t *au, int type)
+{
+    int val, i, llen, tlen, freeloguser=0;
+    int acct_type;
+    pid_t pid;
+    uint16_t taskno;
+    unsigned argc=0, session=0, auid;
+    char *auser = NULL, *loguser, *tty = NULL;
+    char *cmd = NULL, *ausyscall = NULL;
+    char logbuf[240], *logptr, *logbase;
+
+    /* get hostname. */
+    char host[HOST_NAME_MAX];

```

Comment:
>host [](http://example.com/codeflow?start=9&length=4)

Host names (not including the terminating null byte) are limited to HOST_NAME_MAX bytes.
So you need one more char in order to store it as zero terminated string.

Reply:
Fixed, now this code been removed and in the lookup_logname method, allocate HOST_NAME_MAX+1 byte for host name.